Score-based iid sampling

manuelgloeckler opened the pull request:
Completes the missing features based on score estimation #1226. 

- [x] IID interface
- [x] IID util functions
- [x] FNPE
- [x] GAUSS
- [x] JAC
- [x] test

manuelgloeckler commented:
Okey, everything should be implemented now. A few more points:
- [ ] Check if batch jacobian with torch.func.vmap actually works correctly
- [ ] Check if the above or other cause some performance degradation in jac_gauss (although this can be sensitive to how the network is preconditioned)
- [ ] Add an API to pass hyperparameters to the IID method (and make iid_methods more customizable i.e. auto_gauss)
- [ ] Multivariate priors
- [ ] General Empirical prior support for automatic denoising and marginalization (then auto_gauss should become default)

manuelgloeckler commented:
This is now basically done. With the review, one should probably wait until the other score branch and type fixes are merged.

But the major changes are:
- ScoreFnIID classes which manage the score composition
- ScoreUtil, which has a bunch of helpers for "automatic" marginalization and denoising of PyTorch distributions (i.e., what the user can pass as the prior). If there is no analytic solution (or the user does not pass a prior) it will fall back to a rather good MoG approximation.

janfb commented:
@manuelgloeckler #1370 has been merged into `main`. please merge `main` to resolve the conflicts showing up here.

janfb commented on sbi/utils/score_utils.py:
```diff
@@ -0,0 +1,597 @@
+    s: Tensor,
+    x_t: Tensor,
+) -> MixtureSameFamily:
+    """Denoise an empirical distribution.
+
```
please add more details how this is done, i.e., vie gmm etc.

janfb commented on sbi/utils/score_utils.py:
```diff
@@ -0,0 +1,597 @@
+
+
+# Special distributions:
+class UniformNormalPosterior(Distribution):
```
does it make sense to move this to a more general utils? e.g., during linear gaussian testing we need a uniform prior Gaussian likelihood gt posterior.

janfb replied:
please add types throughout - thanks!

manuelgloeckler replied:
I guess. But this is a very constrained "class" i.e. only supports sclar scaling and std per dimension ...

janfb commented on sbi/utils/score_utils.py:
```diff
@@ -0,0 +1,597 @@
+        return torch.where((x >= self.low) & (x <= self.high), lp, -torch.inf)
+
+
+class UniformNormalConvolution(Distribution):
```
same as above, it's great to have this! maybe we can re-use this more generally, e.g,. for the tests? 

Please add types.

janfb commented on tests/linearGaussian_npse_test.py:
```diff
@@ -157,20 +158,22 @@ def simulator(theta):
+@pytest.mark.slow
+@pytest.mark.parametrize("sde_type", ["vp", "ve", "subvp"])
+@pytest.mark.parametrize("with_improper_prior", [False, True])
+def test_npse_iid_inference(sde_type, with_improper_prior):
```
please add docstring to explain these args.

janfb commented on tests/linearGaussian_npse_test.py:
```diff
@@ -179,27 +182,94 @@ def test_npse_iid_inference(num_trials):
+    # This check that our method to handle "general" priors works.
+    # i.e. if NPSE does not get a proper passed by the user.
+    prior_npse = prior if not with_improper_prior else None
+    inference = NPSE(prior_npse, show_progress_bars=True, sde_type=sde_type)
 
     theta = prior.sample((num_simulations,))
     x = linear_gaussian(theta, likelihood_shift, likelihood_cov)
 
     score_estimator = inference.append_simulations(theta, x).train(
-        training_batch_size=100,
+        training_batch_size=200, max_num_epochs=400
     )
```
Can we move this into a fixture to not retrain for every iid method?

Then we can have the iid-methods as `pytest.mark.parametrize` to have more transparency on the test cases. 

We can also have specific num_trials for each iid-method using `pytest.param`

manuelgloeckler replied:
Good idea, not sure how but wil try to do it like that.

janfb commented on tests/linearGaussian_npse_test.py:
```diff
@@ -157,20 +158,22 @@ def simulator(theta):
-    x_o = zeros(num_trials, num_dim)
     num_samples = 1000
     num_simulations = 3000
+    iid_methods = ["fnpe", "gauss", "auto_gauss", "jac_gauss"]
```
have these as pytest params instead (see fixture comment below).

janfb commented on tests/linearGaussian_npse_test.py:
```diff
@@ -157,20 +158,22 @@ def simulator(theta):
     num_samples = 1000
     num_simulations = 3000
+    iid_methods = ["fnpe", "gauss", "auto_gauss", "jac_gauss"]
+    num_good_evals_per_method = {
```
please rename to `num_trials` and use `pytest.param` to have specific num_trials per method.

manuelgloeckler replied:
But the think here is that I also do not want to retrain the score_estimator all the time. If I do it with parameters or fixtures, will it rerun the whole test?

janfb commented on tests/linearGaussian_npse_test.py:
```diff
@@ -179,27 +182,94 @@ def test_npse_iid_inference(num_trials):
-    inference = NPSE(prior, show_progress_bars=True)
+    # This check that our method to handle "general" priors works.
+    # i.e. if NPSE does not get a proper passed by the user.
+    prior_npse = prior if not with_improper_prior else None
```
use an explicit `prior` arg for this test and have params `gaussian` and `None`?

janfb commented on tests/linearGaussian_npse_test.py:
```diff
@@ -179,27 +182,94 @@ def test_npse_iid_inference(num_trials):
 
     score_estimator = inference.append_simulations(theta, x).train(
-        training_batch_size=100,
+        training_batch_size=200, max_num_epochs=400
```
do we still need this increased `num_epochs`? was not this fixed with the recent update on the convergence?

manuelgloeckler replied:
Yeah, will have a look into this. The tests did, however, fail on the new setting (although not by much, I can just change the tolerance of the checks a bit). In general, these will be slightly more sensitive to a small change in approximation performance on 1 trial; it will propagate.

janfb commented on tests/linearGaussian_npse_test.py:
```diff
@@ -179,27 +182,94 @@ def test_npse_iid_inference(num_trials):
+            check_c2st(
+                samples,
+                target_samples,
+                alg=f"npse-vp-gaussian-2D-{iid_method}-{num_trial}iid-trials",
```
add `num_dim` instead of 2D?

janfb commented on tests/linearGaussian_npse_test.py:
```diff
@@ -179,27 +182,94 @@ def test_npse_iid_inference(num_trials):
-    check_c2st(
-        samples, target_samples, alg=f"npse-vp-gaussian-1D-{num_trials}iid-trials"
+
+@pytest.mark.slow
```
this is essentially the same test as above but with `prior=Uniform` no? So I suggest merging the two tests and just using `pytest.mark.parametrize("prior", ("gaussian", "uniform", None))`

or am I missing something?

janfb commented on tests/score_samplers_test.py:
```diff
@@ -8,17 +8,66 @@
+    ],
+)
+@pytest.mark.parametrize("d", [1, 2, 3])
+def test_score_fn_iid_on_different_priors(sde_type, iid_method, d):
```
please add docstring with basic explanation.

janfb commented on tests/score_samplers_test.py:
```diff
@@ -8,17 +8,66 @@
+        "jac_gauss",
+    ],
+)
+@pytest.mark.parametrize("d", [1, 2, 3])
```
`num_dim`

manuelgloeckler commented on sbi/inference/posteriors/score_posterior.py:
```diff
@@ -123,6 +125,9 @@ def sample(
             ts: Time points at which to evaluate the diffusion process. If None, a
                 linear grid between t_max and t_min is used.
+            iid_method: Which method to use for computing the score in the iid setting.
+                We currently support "fnpe", "gauss", "auto_gauss", "jac_gauss".
```
These kinda are the names of the methods, but I can add a bit of detail on what they do and what to care about (from an applied perspective).

manuelgloeckler commented on sbi/inference/posteriors/score_posterior.py:
```diff
@@ -138,7 +143,10 @@ def sample(
         x = self._x_else_default_x(x)
         x = reshape_to_batch_event(x, self.score_estimator.condition_shape)
-        self.potential_fn.set_x(x, x_is_iid=True)
+        is_iid = x.ndim > 1 and x.shape[0] > 1
```
Good catch I think so, too.

manuelgloeckler commented on sbi/inference/posteriors/score_posterior.py:
```diff
@@ -176,6 +184,7 @@ def sample(
     def _sample_via_diffusion(
         self,
         sample_shape: Shape = torch.Size(),
+        x: Optional[Tensor] = None,
```
I did not add that this was part of the #1226 changes. But I can have a look.

manuelgloeckler commented on sbi/inference/posteriors/score_posterior.py:
```diff
@@ -244,6 +253,7 @@ def _sample_via_diffusion(
     def sample_via_ode(
         self,
         sample_shape: Shape = torch.Size(),
+        x: Optional[Tensor] = None,
```
I did not add that this was part of the https://github.com/sbi-dev/sbi/issues/1226 changes. But I can have a look.

manuelgloeckler commented on sbi/inference/potentials/score_based_potential.py:
```diff
@@ -57,7 +58,8 @@ def __init__(
         x_o: Optional[Tensor] = None,
-        iid_method: str = "iid_bridge",
+        iid_method: str = "auto_gauss",
+        iid_params: Optional[dict] = None,
```
Would do Dict[str, Any] can be from floats to tensors to strings.

manuelgloeckler commented on sbi/inference/potentials/score_based_potential.py:
```diff
@@ -172,10 +181,16 @@ def gradient(
+                assert self.prior is not None, "Prior is required for iid methods."
+
+                method_iid = get_iid_method(self.iid_method)
+                # Always creating a new object every call is not efficient...
```
I will delete it. I now cache (using functools) the most expensive operations, which only need to be done once for x_o. So, I think its fine now and  I will just remove the comment (the overhead is now the standard object creation overhead for python).

manuelgloeckler commented on sbi/inference/potentials/score_based_potential.py:
```diff
@@ -217,9 +232,6 @@ def rebuild_flow(
         x_density_estimator = reshape_to_batch_event(
             self.x_o, event_shape=self.score_estimator.condition_shape
         )
-        assert x_density_estimator.shape[0] == 1, (
```
I just removed it, as it was blocking any iid methodology. 

I also think that it is checked in the posterior (and if not we should do it there). The score_fn_iid should anyway reduce this batch dimension to 1 and (atleast) the gauss methods should be applicable to ode sampling (althought not for flow matching).

manuelgloeckler commented on sbi/inference/potentials/score_fn_iid.py:
This is specific to the score-based potential and not necessary for the score-based samplers, as the log_likelihood_over_iid functions work with a bit more effort.

Initially, I put it in samplers/score, but then moved here. It is completely orthogonal to the sampler you decide to use. It just "sums" up the iid dimension in a more complicated way.

manuelgloeckler commented on sbi/samplers/score/correctors.py:
```diff
@@ -63,3 +65,66 @@ def __call__(
+        self.num_steps = num_steps
+
+    def correct(self, theta: Tensor, t0: Tensor, t1: Optional[Tensor] = None) -> Tensor:
+        # TODO: Why is this impacting performance
```
No, can be removed.

manuelgloeckler commented on sbi/samplers/score/diffuser.py:
```diff
@@ -92,17 +92,15 @@ def initialize(self, num_samples: int) -> Tensor:
         # TODO: this fixes the iid setting shape problems, but iid inference via
         # iid_bridge is not accurate.
-        num_batch = self.batch_shape.numel()
+        num_batch = (
+            1 if self.predictor.potential_fn.x_is_iid else self.batch_shape.numel()
+        )
```
yes i.e. if I have 10 iid xs I just want to have 1 sample if I have 10 "batched" xs I also want to have 10 samples.

manuelgloeckler commented on sbi/samplers/score/diffuser.py:
```diff
@@ -92,17 +92,15 @@ def initialize(self, num_samples: int) -> Tensor:
+            1 if self.predictor.potential_fn.x_is_iid else self.batch_shape.numel()
+        )
         init_shape = (num_samples, num_batch) + self.input_shape
         # NOTE: for the IID setting we might need to scale the noise with iid batch
         # size, as in equation (7) in the paper.
+        # NOTE: TODO we need some interface for that
```
This is kinda only relevant for FNPE (and is not that important). I forgot about that, I will have a look but it can be made another issue (I anyway would not recommend using FNPE).

manuelgloeckler replied:
Fixed it. But not perfect

manuelgloeckler commented on sbi/inference/potentials/score_fn_iid.py:
```diff
@@ -0,0 +1,696 @@
+        return score
+
+
+class AbstractGaussCorrectedScoreFn(ScoreFnIID):
```
Agree not a abstract class should be the "Base" class

manuelgloeckler commented:
Thanks for the review. I addressed most of the points above and left open what I wasn't sure about.

Summarizing points:
- Rebased on main.
- Fixed the abstract classes and structure/typing issues.
- Added more documentation to introduced methods.
- Changed the testing - Now the training is a "module" level fixture; hence, it will only be executed once this module is tested.

janfb reviewed (commented):
Thanks for the update @manuelgloeckler, almost done!

Added a couple of final questions and pushed small docs fixes myself.

janfb commented:
all CD tests are passing!

manuelgloeckler commented on sbi/inference/potentials/score_based_potential.py:
```diff
@@ -95,7 +101,9 @@ def set_x(
-        if self._x_o is not None:
+        self.iid_method = iid_method
+        self.iid_params = iid_params
+        if not x_is_iid and (self._x_o is not None):
```
Mhh, I don't think log_prob will work with IID data yet, good point!

 I didn't think about that; the log_prob on IID data must be computed differently than standard i.e. it must build the flow with the "score_fn_iid" and not with the score estimator (or do it with score estimator sum up the log_probs like NLE and substract  (N-1) times the prior log_prob which might be better/easier as I think only the gauss methods have a valid ODE interpretation). Will think about it.

janfb replied:
yes, good catch! So for now we would need to raise an error when calling log_prob given iid data? 
And we move this problem into a feature to be solved during the hackathon?

manuelgloeckler replied:
Yes, lets do that. I will raise an error and add a few comments + create a new issue.

manuelgloeckler commented on sbi/inference/potentials/score_fn_iid.py:
```diff
@@ -0,0 +1,734 @@
+        self.ensure_lam_psd = ensure_lam_psd
+        self.lam_psd_nugget = lam_psd_nugget
+
+    def posterior_precision_est_fn(self, conditions: Tensor) -> Tensor:
```
Mhh, yeah. I think it's best to remove it and replace it with the current "gauss" method. I think it just would be good to have one where the user can provide the "hyperparameter" (maybe even "must").

manuelgloeckler commented on sbi/samplers/score/correctors.py:
```diff
@@ -63,3 +73,74 @@ def __call__(
+        self.std = math.sqrt(2 * self.step_size)
+        self.num_steps = num_steps
+
+    def correct(self, theta: Tensor, t0: Tensor, t1: Optional[Tensor] = None) -> Tensor:
```
Yes, this one does not require it (but the other one needs it hence I made this the signature of all correctors).

manuelgloeckler commented on sbi/inference/potentials/score_fn_iid.py:
```diff
@@ -0,0 +1,734 @@
+        """
+        return torch.ones((1,))
+
+    def marginal_prior(self, time: Tensor, inputs: Tensor) -> Distribution:
```
Can maybe be removed in general (but maybe needed for shape inference). Will check.

manuelgloeckler replied:
I removed it.

janfb reviewed (approved):
great, thanks for the update! 
fixed some typos and will update the changelog to include this into release.
